swift/Ice/async/Server/main.swift

```diff
@@ -0,0 +1,84 @@
+//
+// Copyright (c) ZeroC, Inc. All rights reserved.
+//
+
+import Foundation
+import Ice
+import PromiseKit
+
+class HelloI: Hello {
+    var workQueue: OperationQueue
+    var serialQueue: DispatchQueue // Used for blocking the change of workQueue.isSuspended
+    var concurrentQueue: DispatchQueue // DispatchQueue used in workQueue
+
+    init(_ queue: OperationQueue) {
+        workQueue = queue
+        serialQueue = DispatchQueue(label: "com.zeroc.demo.ice.async.server.serialqueue")
+        concurrentQueue = DispatchQueue(label: "com.zeroc.demo.ice.async.server.workqueue",
+                                        qos: .default,
+                                        attributes: .concurrent)
+        workQueue.underlyingQueue = concurrentQueue
+    }
+
+    func sayHelloAsync(delay: Int32, current _: Current) -> Promise<Void> {
+        return Promise { seal in
+            if delay == 0 {
+                print("Hello World!")
+                seal.fulfill(())
+            } else {
+                serialQueue.sync {
+                    self.workQueue.addOperation {
+                        guard self.workQueue.isSuspended == false else {
```

Comment:
If the `workQueue` is suspended then this operation is added it will never execute. Instead of using `workedQueue.isSuspended` we should just add a `var destroyed: Bool`  and check it instead. That way the operation queue will always execute it's queued items.

Reply:
Not sure we need a second var here, seems simpler to check `isSuspended` before adding to the queue too, so if the queue is already suspended we reject the promise right away, and operations that are already in the queue will throw once they are executed.

Reply:
That's fine too. One question is whether or not `isSuspended` its self is thread safe. If we access it outside of the operation we'll probably have to use an async dispatch.

Reply:
I think we can schedule addOperation asynchronously in the `concurrentQueue`, get rid of the `serialQueue` and set `self.workQueue.isSuspended` from the `concurrentQueue` using a barrier

Reply:
One issue now is that if you add a delayed operation then call shutdown, odds are that the operation queue will be running while destroyed waits. 

So if you do:
```
> d
> d
> d
> s
```

on the server you'll see

```
Shutting down...
Belated Hello World!
Belated Hello World!
Belated Hello World!
```

Reply:
I discussed this with @bernardnormier. It's probably simplest to remove the operation queue and then dispatch asynchronously to the serial queue.